internal/db/inmemory/map.go

```diff
@@ -16,36 +16,64 @@ type Map[K comparable, V Value[K]] struct {
 }
 
 func (m *Map[K, V]) Add(k K, v V) {
-	//TODO implement me
-	panic("implement me")
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.data[k] = v
 }
 
 func (m *Map[K, V]) AddAll(vs []V) {
-	//TODO implement me
-	panic("implement me")
+	m.mu.Lock()
+	for _, v := range vs {
+		m.data[v.Key()] = v
+	}
+	m.mu.Unlock()
 }
 
 func (m *Map[K, V]) Remove(k K) {
-	//TODO implement me
-	panic("implement me")
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.data != nil {
+		delete(m.data, k)
+	}
 }
 
 func (m *Map[K, V]) Get(k K) (V, bool) {
-	//TODO implement me
-	panic("implement me")
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	for keys := range m.data {
+		if keys == k {
+			return m.data[keys], true
+		}
+	}
+	var v V // alternative to *new(V) but without alloc
+	return v, false
 }
 
 func (m *Map[K, V]) Clear() {
-	//TODO implement me
-	panic("implement me")
+	m.mu.Lock()
+	m.data = make(map[K]V)
+	m.mu.Unlock()
 }
 
 func (m *Map[K, V]) Snapshot() []V {
-	//TODO implement me
-	panic("implement me")
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	values := make([]V, 0, len(m.data))
+	for _, value := range m.data {
+		values = append(values, value)
+	}
+	return values
```

Comment:
Iterating a map with range does not guarantee the order in which values are returned. i.e. if you call `Snapshot` twice, you won't get the same list of values in the correct order. 

A `list.List` can maintain the insertion order of items and `Snapshot` can be generated in linear time. For example:

```go
import "container/list"
...

type M struct {
...
 order list.List[K]
}
...
func (m *M) Add(k, v) {
    m.Lock()
    defer m.Unlock()
    m.safedata()[k] = v
    m.order.PushBack(k)
}

func (m *M) Snapshot() []V {
    arr := make([]V, 0, len(m.data))
    // range over iter and add items into arr
}

func (m *M) Iter() iter.Seq[V] {
     return func(yield func(V) bool) {
       m.mu.RLock()
       defer m.mu.RUnlock()
       
       for it := m.order.Front(); it != nil; it = it.Next() {
             if ! yield(m.safedata()[it.Value()]) {
                 return
            }
       }
     }
}
```

Reply:
I wrote this offhand, so it's not syntactically correct, for example, list.List doesn't use generics yet, but I hope you get the idea I'm passing across.